pkg/bootstrap/docker/dockerhelper/filetransfer.go

```diff
@@ -0,0 +1,192 @@
+package dockerhelper
+
+import (
+	"archive/tar"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+type containerDownloader struct {
+	pipeReader *io.PipeReader
+	tarReader  *tar.Reader
+}
+
+func (downloader *containerDownloader) Read(b []byte) (int, error) {
+	return downloader.tarReader.Read(b)
+}
+
+func (downloader *containerDownloader) Close() error {
+	return downloader.pipeReader.Close()
+}
+
+func newContainerDownloader(client *docker.Client, container, path string) *containerDownloader {
+	pipeReader, pipeWriter := io.Pipe()
+	downloader := &containerDownloader{
+		pipeReader: pipeReader,
+		tarReader:  tar.NewReader(pipeReader),
+	}
+
+	go func() {
+		opts := docker.DownloadFromContainerOptions{
+			Path:         path,
+			OutputStream: pipeWriter,
+		}
+		pipeWriter.CloseWithError(client.DownloadFromContainer(container, opts))
+	}()
+
+	return downloader
+}
+
+// StreamFileFromContainer returns an io.ReadCloser from which the contents of a
+// file in a remote container can be read.
+func StreamFileFromContainer(client *docker.Client, container, src string) (io.ReadCloser, error) {
+	downloader := newContainerDownloader(client, container, src)
+
+	for {
+		header, err := downloader.tarReader.Next()
+		if err != nil {
+			return nil, err
+		}
+		if header.Name != filepath.Base(src) || (header.Typeflag != tar.TypeReg && header.Typeflag != tar.TypeRegA) {
+			return nil, errors.New("unexpected tar file content " + header.Name)
+		}
+		return downloader, nil
+	}
+}
+
+// DownloadDirFromContainer downloads an entire directory of files from a remote
+// container.
+func DownloadDirFromContainer(client *docker.Client, container, src, dst string) error {
+	downloader := newContainerDownloader(client, container, src)
+	defer downloader.Close()
```

Comment:
The main reason I didn't was that I'm stripping off the leading directory name when writing files from the incoming tar file, e.g. foo/bar/baz -> bar/baz.  The s2i tar code doesn't currently allow this, and I was hoping to avoid the extract to temporary directory and moving files around alternative.

I could add the relevant functionality into source-to-image and pull it through, but I think it implies bit of refactoring to make it sit well.  Adding a "strip leading directory" flag to ExtractTarStreamWithLogging didn't feel like a great option, and breaking the function apart to abstractions seemed like overkill.  So I made a copy.